dac, deplot pin mutexes; drive motor functions to commands u and l; add drive motor function to the can_reader task

erfi-x2 opened the pull request:
Seems like I added the dac mutex properly, still need some help in figuring out calling the `drive_motor()` function as I don't really understand what parameters to put in place for the arguments `deploy1`, `deploy2`, `motor_ps`, and `dac`.  I assumed I just needed to call `drive_motor()` with 4 values like such: `drive_motor(true, 50, false, 1000)`, but my lsp was giving me an error asking for the 9. Let me know what you think.

wjeffreys96 committed:
WIP: hall sensor and motor drive

erfi-x2 committed:
WIP: added dac mutex, drive motor functions to commands u and l, and added drive motor function to the can_reader task

erfi-x2 committed:
change the mutex to actually be a mutex and added the current argument to the drive motor function

wjeffreys96 commented on controlSystem/RecoveryBoard/firmware-rs/src/bin/drogue.rs:
```diff
@@ -123,12 +123,12 @@ static UMB_ON_MTX: UmbOnType = Mutex::new(None);
 static SYSTEM_STATE_MTX: Mutex<ThreadModeRawMutex, Option<DrogueState>> = Mutex::new(None);
-
 static RING_POSITION_CHANNEL: Channel<ThreadModeRawMutex, RingPosition, 5> = Channel::new();
+static DAC_MTX: Mutex<ThreadModeRawMutex, Option<Dac<'static, DAC1,Async>>, 5> = Mutex::new(None);
```
Mutex does not need the third generic argument here ("5"), as it only holds one value at any given time.

erfi-x2 replied:
Fixed!

wjeffreys96 commented on controlSystem/RecoveryBoard/firmware-rs/src/bin/drogue.rs:
```diff
@@ -190,6 +190,7 @@ async fn main(spawner: Spawner) {
         *(ADC_MTX.lock().await) = Some(adc);
         *(UMB_ON_MTX.lock().await) = Some(umb_on);
         *(SYSTEM_STATE_MTX.lock().await) = Some(sys_state);
+        *(DAC_MTX.lock().await = Some(dac));
```
Parentheses should only be around the left side of the assignment. The purpose is to de-reference the mutex with the asterisk and assign to the inner value (what the mutex holds).

erfi-x2 replied:
Fixed!

wjeffreys96 commented on controlSystem/RecoveryBoard/firmware-rs/src/bin/drogue.rs:
```diff
@@ -266,9 +268,11 @@ pub async fn cli(uart: BufferedUart<'static>) {
                 }
                 "l" => {
-                    // TODO: implement
+                    drive_motor(PB4, PB5, &PB6, true, 50, false, 1000, RingPosition, &DAC1);
```
drive_motor needs to take in the mutex. How it is, you are passing a reference to a type. Be careful with this, because embassy has several modules that can refer to the same peripheral. There is the pac (peripheral access crate, for unsafe things), the dac module (the actual peripheral module), and the peripherals module (types for peripherals), and possibly more depending on what were talking about.

The problem with the pins is the same. These are types for the pins. We need to pass the pins into the task that calls drive_motor, and then pass those into drive motor. This is what I was talking about with prop chains, (React terminology) where an argument gets passed into something, which then gets passed into another thing as an argument and so on. They can get hard to manage, so if you have any ideas how to simplify that would be good. However, if this is as deep as the chain goes it's probably fine.

erfi-x2 replied:
I think this should plus the can_reader is as far as it goes. Is there a way to pass the pins from `main()` to each task we want? As it is now the code:


   `let deploy_1 = Output::new(p.PB4, Level::Low, Speed::Medium);`
   `let deploy_2 = Output::new(p.PB5, Level::Low, Speed::Medium);`
   `let motor_ps = Output::new(p.PB6, Level::High, Speed::Medium);`


is still is unused in `main()`. I would think we could just define those in both the `cli()` task and the `can_reader()` task, is there any problems we would run into doing so?

erfi-x2 replied:
For the arguments of `RingPosition` and the dac, I am a bit lost on how to deal with them, you help with them would be great.

wjeffreys96 replied:
That would not work. The executor needs to spawn the tasks inside main(). That is where the pins get passed in, since our references to them are created with the HAL we create (p) by calling embassy_stm32::init(). Look to how the other tasks are spawned and the associated functions with the task decorator to see how this all works.

wjeffreys96 commented on controlSystem/RecoveryBoard/firmware-rs/src/bin/drogue.rs:
```diff
@@ -266,9 +268,11 @@ pub async fn cli(uart: BufferedUart<'static>) {
+                    drive_motor(PB4, PB5, &PB6, true, 50, false, 1000, RingPosition, &DAC1);
+                }
+                "u" => {
+                    drive_motor(PB4, PB5, PB6, false, 50, false, 1000, RingPosition, &DAC1);
```
Same issue as above here

wjeffreys96 commented on controlSystem/RecoveryBoard/firmware-rs/src/bin/drogue.rs:
```diff
@@ -346,7 +350,6 @@ fn get_ring_position(sensor1_state: SensorState, sensor2_state: SensorState) ->
 }
 
 async fn limit_motor_current(mut dac: Dac<'static, DAC1, Async>, ma: u16) {
```
We can't pass in the actual dac anymore. It is now inside the mutex and we must change the function signature to reflect this. Creating a custom type for the mutex like I have for the others really eases this process. When we use the dac, we must lock the mutex with let mut dac_unlocked = DAC_MTX.lock().await and then use it with if let Some(dac) = dac_unlocked.as_mut() {...}. be careful with how long we have the mutex locked for, and control it with a scope ("{}").

erfi-x2 replied:
Check out the new commit, I believe it should be correct but I don't trust myself.

wjeffreys96 commented on controlSystem/RecoveryBoard/firmware-rs/src/bin/drogue.rs:
```diff
@@ -373,7 +375,7 @@ async fn drive_motor(
 
     if lock_mode {
         deploy2.set_high();
-        limit_motor_current(dac, 1000).await;
```
Same as above

wjeffreys96 commented on controlSystem/RecoveryBoard/firmware-rs/src/bin/drogue.rs:
```diff
@@ -426,7 +428,8 @@ pub async fn read_hall_sensor(
 
 #[embassy_executor::task]
-async fn can_reader(mut can_rx: CanRx<'static>, mut can: Can<'static>) -> () {
+async fn can_reader(mut can_rx: CanRx<'static>, mut can: Can<'static>, can_id: u16) -> () {
```
We don't need to pass in a can id. Any id we plan on matching should be imported as a constant from the can.rs file and will be in scope for this whole file.

erfi-x2 replied:
Gotcha, should be imported in the latest code now. I am still iffy on how we would receive the can_rx contents as u16 in order to check for the values of the can_id.

wjeffreys96 reviewed (changes requested):
Let me know if you need any help addressing these requested changes.

erfi-x2 committed:
changed the dac mutex to a type and applied appropriate fixes

erfi-x2 commented:
At any point you want to take over the code, feel free to do so.

erfi-x2 committed:
try to pass to tasks and changed logic to match id

It is in a wrong, non-working state. Can't get the pins to pass onto
each task properly and having mismatched types on the check for matching
the ids.

Commit:
fix all errors

wjeffreys96 committed:
Merge branch 'ers-embassy' into ers-embassy-hess
